File: 155-最小栈.py
"""
设计一个支持 push，pop，top 操作，并能在常数时间内检索到最小元素的栈。

push(x) -- 将元素 x 推入栈中。
pop() -- 删除栈顶的元素。
top() -- 获取栈顶元素。
getMin() -- 检索栈中的最小元素。
示例:

MinStack minStack = new MinStack();
minStack.push(-2);
minStack.push(0);
minStack.push(-3);
minStack.getMin();   --> 返回 -3.
minStack.pop();
minStack.top();      --> 返回 0.
minStack.getMin();   --> 返回 -2.
"""

class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.mystack = []
        self.minValue = None
        

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        self.mystack.append(x)
        if self.minValue == None or self.minValue > x:
            self.minValue = x

        

    def pop(self):
        """
        :rtype: void
        """
        popItem = self.mystack.pop()
        if len(self.mystack) == 0:
            self.minValue = None
            return popItem
        
        if popItem == self.minValue:
            self.minValue = self.mystack[0]
            for v in self.mystack:
                if v < self.minValue:
                    self.minValue = v
        return popItem
        

    def top(self):
        """
        :rtype: int
        """
        return self.mystack[-1]
        

    def getMin(self):
        """
        :rtype: int
        """
        return self.minValue

File: LoveSwiftLeetcode/155.最小栈.swift
/*
 * @lc app=leetcode.cn id=155 lang=swift
 *
 * [155] 最小栈
 */

// @lc code=start

class MinStack {
    var stack = [Int]()
    var minStack = [Int.max]
    init() {

    }
    
    func push(_ val: Int) {
        stack.append(val)
        let cnt = minStack.count
        minStack.append(min(minStack[cnt - 1], val))
    }
    
    func pop() {
        if !stack.isEmpty {
            let _ = stack.popLast()
            let _ = minStack.popLast()
        }
    }
    
    func top() -> Int {
        if stack.isEmpty == false {
            return stack.last!
        }
        return -1
    }
    
    func getMin() -> Int {
        if stack.isEmpty == false {
            return minStack.last!
        }
        return -1
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * let obj = MinStack()
 * obj.push(val)
 * obj.pop()
 * let ret_3: Int = obj.top()
 * let ret_4: Int = obj.getMin()
 */
// @lc code=end

